### PR TITLE
do not create dest on unreadable src

### PR DESCRIPTION
**Pull request by schnittstabil**
IMO that would fix the second issue of https://github.com/sindresorhus/cp-file/pull/4#issuecomment-75506946.
@kevva: what do you think?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,3 @@
 node_modules
+tmp
```
Comment (kevva) [inline .gitignore]: Remove these. It's just a sign that something isn't working.

File: index.js [hunk 2]
```diff
@@ -14,59 +15,47 @@ module.exports = function (src, dest, opts, cb) {
...
-			var read = fs.createReadStream(src);
-			var write = fs.createWriteStream(dest);
+			var write = fs.createWriteStream(dest,
+				{flags: opts.overwrite ? 'w' : 'wx'});
+
+			write.on('error', function (err) {
+				if (!opts.overwrite && err.code === 'EEXIST') {
+					cb();
+					return;
+				}
```
Comment (kevva) [inline index.js]:
```js
}

cb();
```

File: index.js [hunk 3]
```diff
@@ -14,59 +15,47 @@ module.exports = function (src, dest, opts, cb) {
...
+				cb(err);
```
Comment (kevva) [inline index.js]:
```js
}

cb();
```

File: index.js [hunk 4]
```diff
@@ -88,20 +78,24 @@ module.exports.sync = function (src, dest, opts) {
...
 
-	var BUF_LENGTH = 100 * 1024;
-	var buf = new Buffer(BUF_LENGTH);
-	var read = fs.openSync(src, 'r');
-	var write = fs.openSync(dest, 'w');
-	var stat = fs.fstatSync(read);
-	var bytesRead = BUF_LENGTH;
-	var pos = 0;
+	var write;
+	try {
```
Comment (kevva) [inline index.js]:
```js
var write;

try {
```

File: index.js [hunk 5]
```diff
@@ -88,20 +78,24 @@ module.exports.sync = function (src, dest, opts) {
...
+		write = fs.openSync(dest, opts.overwrite ? 'w' : 'wx');
+	} catch (err) {
+		if (!opts.overwrite && err.code === 'EEXIST') {
+			return;
+		}
+	}
+	fs.writeSync(write, buf, 0, bytesRead);
```
Comment (kevva) [inline index.js]:
```js
}

fs.writeSync(write, buf, 0, bytesRead);
```

File: index.js [hunk 6]
```diff
@@ -88,20 +78,24 @@ module.exports.sync = function (src, dest, opts) {
...
 
+	var pos = bytesRead;
```
Comment (kevva) [inline index.js]: This could be moved further up, right?

File: test.js [hunk 7]
```diff
@@ -17,10 +17,23 @@ function assertDateEqual(actual, expected, message) {
...
+	[
+		'tmp',
+		'EMPTY',
+		'subdir/tmp',
+	].forEach(function (file) {
+		try {
+			fs.unlinkSync(file);
+		} catch (err) {}
+	});
+	[
```
Comment (kevva) [inline test.js]:
```js
});

[
```

File: index.js [hunk 8]
```diff
@@ -14,59 +15,47 @@ module.exports = function (src, dest, opts, cb) {
...
 	}
 
-	cb = cb || function () {};
+	cb = onetime(cb) || function () {};
 	opts = objectAssign({overwrite: true}, opts);
 
-	var cbCalled = false;
+	var read = fs.createReadStream(src);
+	var readListener = onetime(startWrite);
+	read.on('error', cb);
```
Comment (kevva) [inline index.js]:
```js
var readListener = onetime(startWrite);

read.on('error', cb);
```

File: index.js [hunk 9]
```diff
@@ -14,59 +15,47 @@ module.exports = function (src, dest, opts, cb) {
...
 		mkdirp(path.dirname(dest), function (err) {
 			if (err && err.code !== 'EEXIST') {
 				cb(err);
 				return;
 			}
 
-			var read = fs.createReadStream(src);
-			var write = fs.createWriteStream(dest);
+			var write = fs.createWriteStream(dest,
+				{flags: opts.overwrite ? 'w' : 'wx'});
```
Comment (kevva) [inline index.js]:
```js
var write = fs.createWriteStream, dest {flags: opts.overwrite ? 'w' : 'wx'});
```

File: index.js [hunk 10]
```diff
@@ -88,20 +78,24 @@ module.exports.sync = function (src, dest, opts) {
...
-	var buf = new Buffer(BUF_LENGTH);
-	var read = fs.openSync(src, 'r');
-	var write = fs.openSync(dest, 'w');
-	var stat = fs.fstatSync(read);
-	var bytesRead = BUF_LENGTH;
-	var pos = 0;
+	var write;
+	try {
+		write = fs.openSync(dest, opts.overwrite ? 'w' : 'wx');
+	} catch (err) {
```
Comment (kevva) [inline index.js]: The error should be handled in some way. `throw` it or something.

File: index.js [hunk 11]
```diff
@@ -14,59 +15,47 @@ module.exports = function (src, dest, opts, cb) {
 		opts = {};
 	}
 
-	cb = cb || function () {};
+	cb = onetime(cb) || function () {};
```
Comment (kevva) [inline index.js]: Shouldn't it be `onetime(cb || function () {});`?

**Comment by kevva** [comment]
So should this be merged in before mine (#4)?

File: .gitignore (hunk 1 above)
Comment (schnittstabil) [inline .gitignore]: I've substituted  `afterEach` with `beforeEach` in `test.js`:
```diff
- afterEach(function () {
-	try {
-		fs.unlinkSync('tmp');
-	} catch (err) {}
+beforeEach(function () {
```
This allows one to run `mocha --bail` and examine the files created by a failing test. So I think ignoring them would be useful.

File: index.js (hunk 6 above)
Comment (schnittstabil) [inline index.js]: Yes, but it's only used within the loop directly below.

Comment (kevva) [inline index.js]: Easier to read if as many `var` as possible is declared at the same place imo.

File: index.js (hunk 11 above)
Comment (schnittstabil) [inline index.js]: That's almost the same, it only differs if `cb` is `undefined`: my would be negligible faster.

**Comment by schnittstabil** [comment]
It would be instead of #4, the PR prevents the file creation observed by you. But throws an Error if `dest` is a directory, which your version would suppress.

File: index.js (hunk 11 above)
Comment (schnittstabil) [inline index.js]: Oh damn, you are right :-)

File: .gitignore (hunk 1 above)
Comment (schnittstabil) [inline .gitignore]: You're right. I think I've found a better way to support the `mocha --bail` workflow.

**Comment by schnittstabil** [comment]
@kevva: Thank you for your suggestions.
@sindresorhus, @kevva: if you have no objections, I'll squash and merge this into master.

**Comment by sindresorhus** [comment]
:+1:

**Comment by kevva** [comment]
Sounds good to me!

**Pull request closed without merging**